java/arcs/core/storage/api/EntityBase.kt

```diff
@@ -0,0 +1,206 @@
+
+    /** Checks that the given value is of the expected type. */
+    private fun checkType(field: String, value: Any?, type: FieldType) = when (type) {
+        is FieldType.Primitive -> when (type.primitiveType) {
+            PrimitiveType.Boolean -> require(value is Boolean) {
+                "Expected Boolean for $entityClassName.$field, but received $value."
+            }
+            PrimitiveType.Number -> require(value is Double) {
+                "Expected Double for $entityClassName.$field, but received $value."
+            }
+            PrimitiveType.Text -> require(value is String) {
+                "Expected String for $entityClassName.$field, but received $value."
+            }
+        }
+        is FieldType.EntityRef -> {
+            // TODO: Handle References.
+            throw NotImplementedError("References aren't supported yet.")
+        }
+    }
+
+    /** Combines all singleton/collection field data into a single map. */
+    private fun allFields(): Map<String, Any?> = mutableMapOf<String, Any?>().apply {
+        putAll(singletons)
+        putAll(collections)
+    }
+
+    final override fun reset() {
+        schema.fields.singletons.keys.forEach { singletons[it] = null }
+        schema.fields.collections.keys.forEach { collections[it] = emptySet() }
+    }
+
+    override fun serialize() = RawEntity(
+        id = entityId ?: NO_REFERENCE_ID,
+        singletons = singletons.mapValues { (field, value) ->
+            value?.let { toReferencable(it, getSingletonType(field)) }
+        },
+        collections = collections.mapValues { (field, values) ->
+            val type = getCollectionType(field)
+            values.map { toReferencable(it, type) }.toSet()
+        }
+    )
+
+    /**
+     * Populates the entity from the given [RawEntity] serialization. Must only be called on a
+     * fresh, empty instance.
+     */
+    protected fun deserialize(rawEntity: RawEntity) {
+        _entityId = if (rawEntity.id == NO_REFERENCE_ID) null else rawEntity.id
+        rawEntity.singletons.forEach { (field, value) ->
+            setSingletonValue(field, value?.let { fromReferencable(it, getSingletonType(field)) })
+        }
+        rawEntity.collections.forEach { (field, values) ->
+            val type = getCollectionType(field)
+            setCollectionValue(field, values.map { fromReferencable(it, type) }.toSet())
+        }
+    }
+
+    override fun ensureIdentified(idGenerator: Id.Generator, handleName: String) {
+        if (_entityId == null) {
+            _entityId = idGenerator.newChildId(
+                // TODO: should we allow this to be plumbed through?
+                idGenerator.newArcId("dummy-arc"),
+                handleName
+            ).toString()
+        }
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is EntityBase) return false
+        if (entityClassName != other.entityClassName) return false
```

Comment:
Hypothetical: If two RawEntities with the same schemaHash, but code-genned into two different classes have the same entityId, or equivalent rawentity fields should they be equal?

e.g. user makes schema1.arcs in foo/bar package, and copy-pastes schema1.arcs into foo/baz package. They code-gen the same schema into two separate Java packages, but the underlying schema and data is the same. 

This could be especially an issue if there are lots of inline schemas being used, which could have the same schemaHash and fields, but being code-gen duplicated many times?

Reply:
That's a really good point, I hadn't thought about that. Fortunately I think the code is already correct, but I'll add a test case to ensure that. The generated subclasses do not override the equals method, and the EntityBase equals method doesn't check that the subclasses match, so this should just work.